### ISSUE TITLE
fill_null silently change pl.Array to pl.List

### ISSUE THREAD

**Issue by Sage0614** [pola-rs/polars#20327]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
import numpy as np

before = pl.DataFrame(
    {"x": [[1, 2, 3], [4, 5, 6], None], "y": ["a", "b", "c"]},
    schema=pl.Schema({"x": pl.Array(pl.Int64, shape=3), "y": pl.Utf8}),
)

print(before)
after = before.with_columns(pl.col("x").fill_null([0, 0, 0]))
print(after)
```

### Log output

```shell
shape: (3, 2)
┌───────────────┬─────┐
│ x             ┆ y   │
│ ---           ┆ --- │
│ array[i64, 3] ┆ str │
╞═══════════════╪═════╡
│ [1, 2, 3]     ┆ a   │
│ [4, 5, 6]     ┆ b   │
│ null          ┆ c   │
└───────────────┴─────┘
shape: (3, 2)
┌───────────┬─────┐
│ x         ┆ y   │
│ ---       ┆ --- │
│ list[i64] ┆ str │
╞═══════════╪═════╡
│ [1, 2, 3] ┆ a   │
│ [4, 5, 6] ┆ b   │
│ [0, 0, 0] ┆ c   │
└───────────┴─────┘
```

### Issue description

the filled column type change from pl.Array to pl.List scilently, I don't expect Array to cast to List scilently as it is potentially dengerous.
also currently I don't find a way to fill an array column with array and keep it as array, none of fill_null with [0,0,0], (0,0,0), np.array([0,0,0]) works as expected

### Expected behavior

┌───────────┬─────┐
│ x         ┆ y   │
│ ---       ┆ --- │
│ array[i64,3] ┆ str │
╞═══════════╪═════╡
│ [1, 2, 3] ┆ a   │
│ [4, 5, 6] ┆ b   │
│ [0, 0, 0] ┆ c   │
└───────────┴─────┘

### Installed versions

<details>

```
--------Version info---------
Polars:              1.17.1
Index type:          UInt32
Platform:            Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.35
Python:              3.11.4 (main, Aug 24 2023, 11:18:03) [GCC 11.4.0]
LTS CPU:             False

----Optional dependencies----
adbc_driver_manager  <not installed>
altair               <not installed>
boto3                1.28.42
cloudpickle          <not installed>
connectorx           <not installed>
deltalake            <not installed>
fastexcel            <not installed>
fsspec               2023.6.0
gevent               <not installed>
google.auth          2.23.3
great_tables         <not installed>
matplotlib           3.7.2
nest_asyncio         1.5.7
numpy                1.25.2
openpyxl             <not installed>
pandas               2.0.3
pyarrow              13.0.0
pydantic             1.10.12
pyiceberg            <not installed>
sqlalchemy           2.0.20
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

</details>

**Comment by cmdlineluser** [comment]
I think there is a general issue with fill_null not being strict by default: https://github.com/pola-rs/polars/issues/13789

You can use `pl.lit` to specify an Array dtype.

```python
before.with_columns(
    pl.col("x").fill_null(pl.lit([0, 0, 0], dtype=pl.Array(int, 3)))
)

# shape: (3, 2)
# ┌───────────────┬─────┐
# │ x             ┆ y   │
# │ ---           ┆ --- │
# │ array[i64, 3] ┆ str │
# ╞═══════════════╪═════╡
# │ [1, 2, 3]     ┆ a   │
# │ [4, 5, 6]     ┆ b   │
# │ [0, 0, 0]     ┆ c   │
# └───────────────┴─────┘
```

**Comment by miccoli** [comment]
Let me add a comment to clarify the issue:

1. `pl.lit([0, 0, 0])` is by default a `List(Int64)`
2.  `fill_null` by design silently upcasts to the common supertype, which indeed is `List(Int64)`

So even if `fill_null` is made strict by default, 
```
before.with_columns(
    pl.col("x").fill_null([0, 0, 0])
)
```
will still fail.

Note also that the correct way of specifying a single `Array` row with numpy is `np.array([[0, 0, 0]])`, so that 
```python
before.with_columns(
    pl.col("x").fill_null(np.array([[0, 0, 0]]))
)

# shape: (3, 2)
# ┌───────────────┬─────┐
# │ x             ┆ y   │
# │ ---           ┆ --- │
# │ array[i64, 3] ┆ str │
# ╞═══════════════╪═════╡
# │ [1, 2, 3]     ┆ a   │
# │ [4, 5, 6]     ┆ b   │
# │ [0, 0, 0]     ┆ c   │
# └───────────────┴─────┘
```
works as expected.

**Comment by Sage0614** [comment]
> Let me add a comment to clarify the issue:
> 
> 1. `pl.lit([0, 0, 0])` is by default a `List(Int64)`
> 2. `fill_null` by design silently upcasts to the common supertype, which indeed is `List(Int64)`
> 
> So even if `fill_null` is made strict by default,
> 
> ```
> before.with_columns(
>     pl.col("x").fill_null([0, 0, 0])
> )
> ```
> 
> will still fail.
> 
> Note also that the correct way of specifying a single `Array` row with numpy is `np.array([[0, 0, 0]])`, so that
> 
> ```python
> before.with_columns(
>     pl.col("x").fill_null(np.array([[0, 0, 0]]))
> )
> 
> # shape: (3, 2)
> # ┌───────────────┬─────┐
> # │ x             ┆ y   │
> # │ ---           ┆ --- │
> # │ array[i64, 3] ┆ str │
> # ╞═══════════════╪═════╡
> # │ [1, 2, 3]     ┆ a   │
> # │ [4, 5, 6]     ┆ b   │
> # │ [0, 0, 0]     ┆ c   │
> # └───────────────┴─────┘
> ```
> 
> works as expected.

Ah, this is really confusing about what the array expect, to me if the fill_null( ) api take str input for a pl.String column and take a  int input for a pl.Int column, it should take an array of shape = 3 for an array [i64,3] column, not array [i64,(1,3)], as they mean different thing in numpy, this might be a separate ticket and should be explictly documented.

**Comment by miccoli** [comment]
> Ah, this is really confusing about what the array expect, to me if the fill_null( ) api take str input for a pl.String column and take a int input for a pl.Int column, it should take an array of shape = 3 for an array [i64,3] column, not array [i64,(1,3)], as they mean different thing in numpy, this might be a separate ticket and should be explictly documented.

I agree: I'm not a polars developer and I'm learning by trial and error too.

The confusing point is that `fill_null( )` casting/promotion rules are not clearly spelled out in the docs..., but it seems that there is no consensus on this point, according to #13789.

As what regards numpy arrays, you need to realise that actually the numpy to polars mapping works at the `pl.Series` level. The leading dimension of the numpy array is the Series shape (a singleton) and the remaining dimensions are mapped to the array. Try the following examples:
```
pl.Series(np.arange(6))
#
# shape: (6,)
# Series: '' [i64]
# [
# 	0
# 	1
# 	2
# 	3
# 	4
# 	5
# ]
# 
pl.Series(np.arange(6).reshape((3,2,1))
#
# shape: (3,)
# Series: '' [array[i64, (2, 1)]]
# [
# 	[[0], [1]]
# 	[[2], [3]]
# 	[[4], [5]]
# ]
```
Therefore for obtaining a single row of type `pl.Array` you need a leading `1` dimension, and this explains the (strange) need for `np.array([[0, 0, 0]])` in `fill_null`.
Should `fill_null` provide the leading singleton `(1,)`?

Finally instead of the cryptic `np.array([[0, 0, 0]])` , it is possible to explicitly state that we add a new singleton dimension using `np.newaxis` 
```
before.with_columns(
    pl.col("x")
    .fill_null(np.array([0, 0, 0])[np.newaxis, ...])
)
```
